bndtools.core/resources/unprocessed/serviceTemplates/java-service-consumer/$srcDir$/$basePackageDir$/$serviceName$Consumer.java

```diff
@@ -0,0 +1,40 @@
+package $basePackageName$;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import $api_package$.$serviceName$Service;
+
+@Component(immediate=true)
+public class $serviceName$Consumer {
+	
+	private $serviceName$Service service;
+	
+	/**
+	 * The bind and unbind methods will be called by the service component runtime
+	 * when a $serviceName$Service reference is detected in the local service registry
+	 * @param s
+	 */
+	@Reference
+	void bind$serviceName$Service($serviceName$Service s) {
+		this.service = s;
+		System.out.println("$serviceName$Service instance bound to consumer");
+	}
+	
+	void unbind$serviceName$Service($serviceName$Service s) {
+		this.service = null;
+		System.out.println("$serviceName$Service instance unbound from consumer");
+	}
+	
+	/**
+	 * The activate method will be called immediately upon the $serviceName$Service 
+	 * reference being bound to this component via the bind method$serviceName$Service
+	 * method.
+	 */
+	void activate() {
```

Comment:
Hmm, you're using a default in DS which I never really liked. I also think using the constructor activation far superior since you can make the instance fields final. At least add an @Activate ann.

Reply:
I just forgot the @activate annotation.  I'll add that.   I didn't want to pick the most elegant default...but rather one that wouldn't require much to understand it's behavior (e.g. explicit bind/unbind, possible generalizations like dynamics, etc).  I haven't honestly kept up on all the injection and annotation methods for most recent DS.   Could you point me to an example (from spec or elsewhere) of what you would prefer?

Reply:
`@Activate`

The coolest is constructor injection. It allows you to make your fields final, I've fallen heads over heels in love iwth it. Wonder how I could've lived without! :-)

     
     final $serviceName$Service service;
     @Activate
     public $serviceName$Consumer(@Reference $serviceName$Service service) {
        this.service = service;
     }

I did not remove the `private` by accident, in almost every test situation you want to be able to access it. I am a strong believer in package private. The advantage of private is that you keep your hands free in the future are also valid for package private. And it is less of an eye sore! :-)

Reply:
For my information:   How does constructor injection work if the service is dynamic?